libraries/WebServer/src/Parsing.cpp

```diff
@@ -304,12 +304,39 @@ void WebServer::_uploadWriteByte(uint8_t b){
 
 int WebServer::_uploadReadByte(WiFiClient& client){
   int res = client.read();
-  if(res == -1){
-    while(!client.available() && client.connected())
-      delay(2);
-    res = client.read();
+  if(res < 0) {
+    // keep trying until you either read a valid byte or timeout
+    const unsigned long startMillis = millis();
+    const long timeoutIntervalMillis = HTTP_MAX_POST_WAIT;
+    unsigned long currentMillis = startMillis;
+    for(;;) {
+      // loosely modeled after blinkWithoutDelay pattern
+      while(!client.available() && client.connected()){
+        delay(2);
+      }
+
+      res = client.read();
+      if(res >= 0) {
+        return res; // exit on a valid read
+      }
+      // NOTE: it is possible to get here and have all of the following
+      //       assertions hold true
+      //
+      //       -- client.available() > 0
+      //       -- client.connected == true
+      //       -- res == -1
+      //
+      //       a simple retry strategy overcomes this which is to say the
+      //       assertion is not permanent, but the reason that this works
+      //       is ellusive, and possibly indicative of a more subtle underlying
+      //       issue
+
+      currentMillis = millis();
+      if(currentMillis - startMillis >= timeoutIntervalMillis) {
+        return -1; // exit on a timeout
+      }
+    }
   }
-  return res;
```

Comment:
ALWAYS return as last line ;) compilers complain.

Reply:
ack, I updated this on my branch, it was actually a bug